lms/djangoapps/mobile_api/test_middleware.py

```diff
@@ -0,0 +1,112 @@
+        fake_response = HttpResponse()
+        response = self.middleware.process_response(fake_request, fake_response)
+        self.assertEquals(200, response.status_code)
+        with self.assertRaises(KeyError):
+            response[AppVersionUpgrade.LATEST_VERSION]
+        with self.assertRaises(KeyError):
+            response[AppVersionUpgrade.UPGRADE_DEADLINE]
+
+    @ddt.data(
+        "edX/org.edx.mobile (5.5.5; OS Version 9.2 (Build 13C75))",
+        "Dalvik/2.1.0 (Linux; U; Android 5.1; Nexus 5 Build/LMY47I) edX/org.edx.mobile/5.5.5",
+    )
+    def test_no_update(self, user_agent):
+        """ user has the latest app version """
+        fake_request = HttpRequest()
+        fake_request.META['HTTP_USER_AGENT'] = user_agent
+        self.assertIsNone(self.middleware.process_request(fake_request))
+        fake_response = HttpResponse()
+        response = self.middleware.process_response(fake_request, fake_response)
+        self.assertEquals(200, response.status_code)
+        with self.assertRaises(KeyError):
+            response[AppVersionUpgrade.LATEST_VERSION]
+        with self.assertRaises(KeyError):
+            response[AppVersionUpgrade.UPGRADE_DEADLINE]
+
+    @ddt.data(
+        "edX/org.edx.mobile (1.0.0; OS Version 9.2 (Build 13C75))",
+        "edX/org.edx.mobile (2.2.1; OS Version 9.2 (Build 13C75))",
+        "edX/org.edx.mobile (2.2.1.RC; OS Version 9.2 (Build 13C75))",
+        "Dalvik/2.1.0 (Linux; U; Android 5.1; Nexus 5 Build/LMY47I) edX/org.edx.mobile/1.0.0",
+        "Dalvik/2.1.0 (Linux; U; Android 5.1; Nexus 5 Build/LMY47I) edX/org.edx.mobile/2.2.1",
+        "Dalvik/2.1.0 (Linux; U; Android 5.1; Nexus 5 Build/LMY47I) edX/org.edx.mobile/2.2.1.RC",
+    )
+    def test_upgrade_required(self, user_agent):
+        """ when;
+        1. user version < min supported version
+        2. user version >= min supported version and user version < next min supported version
+        and request timestamp <= update required timestamp
+        """
+        fake_request = HttpRequest()
+        fake_request.META['HTTP_USER_AGENT'] = user_agent
+        result = self.middleware.process_request(fake_request)
+        self.assertIsNotNone(result)
+        response = self.middleware.process_response(fake_request, result)
+        self.assertEquals(426, response.status_code)
+        self.assertEqual("5.5.5", response[AppVersionUpgrade.LATEST_VERSION])
+        self.assertEqual('2016-02-22 00:00:00+00:00', response[AppVersionUpgrade.UPGRADE_DEADLINE])
+
+    @ddt.data(
+        "edX/org.edx.mobile (4.1.1; OS Version 9.2 (Build 13C75))",
+        "edX/org.edx.mobile (4.2.2.RC; OS Version 9.2 (Build 13C75))",
+        "Dalvik/2.1.0 (Linux; U; Android 5.1; Nexus 5 Build/LMY47I) edX/org.edx.mobile/4.1.1",
+        "Dalvik/2.1.0 (Linux; U; Android 5.1; Nexus 5 Build/LMY47I) edX/org.edx.mobile/4.2.2.RC",
+    )
+    def test_new_update_available(self, user_agent):
```

Comment:
Can we re-order the tests to:
```
test_version_is_unsupported()
test_version_is_deprecated()
test_version_is_not_latest()
test_version_is_latest()
```

Also the test for the second case needs to be added.